# Movies Tonight Part 3: SQL DML

In [2]:
%load_ext sql
%sql sqlite:///Movies.db

'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

In [3]:
%%sql 
SELECT * FROM DATASET LIMIT 10;
PRAGMA table_info(DATASET)

 * sqlite:///Movies.db
Done.
Done.


[(0, 'TName', 'TEXT', 0, None, 0),
 (1, 'Location', 'TEXT', 0, None, 0),
 (2, 'Phone', 'TEXT', 0, None, 0),
 (3, 'MTitle', 'TEXT', 0, None, 0),
 (4, 'ShowTime', 'TEXT', 0, None, 0),
 (5, 'Rating', 'TEXT', 0, None, 0),
 (6, 'CCode', 'TEXT', 0, None, 0),
 (7, 'CName', 'TEXT', 0, None, 0)]

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [4]:
%%sql
DELETE FROM MOVIES;
INSERT INTO MOVIES(TITLE, RATING)
    SELECT DISTINCT MTitle, Rating
    FROM DATASET;
SELECT *
FROM MOVIES;

DELETE FROM THEATERS;
INSERT INTO THEATERS(NAME, LOCATION, PHONE)
    SELECT DISTINCT TName, Location, Phone
    FROM DATASET;
SELECT *
FROM THEATERS;

DELETE FROM ARTISTS;
INSERT INTO ARTISTS(NAME)
    SELECT DISTINCT CName
    FROM DATASET;
SELECT * 
FROM ARTISTS;

DELETE FROM SHOWS;
INSERT INTO SHOWS(SHOWTIME, MID, TID)
    SELECT DISTINCT ShowTime, MID, TID
    FROM DATASET
        JOIN MOVIES ON (MOVIES.TITLE=DATASET.MTitle)
        JOIN THEATERS ON (THEATERS.NAME=DATASET.TName);
SELECT SHOWTIME, TITLE, NAME 
FROM SHOWS 
    JOIN MOVIES USING (MID) 
    JOIN THEATERS USING (TID);
    
DELETE FROM CREDITS;
INSERT INTO CREDITS(CCODE, AID, MID)
SELECT DISTINCT CCode, AID, MID
FROM DATASET
    JOIN ARTISTS ON (ARTISTS.NAME=DATASET.CName)
    JOIN MOVIES ON (MOVIES.TITLE=DATASET.MTitle);


 * sqlite:///Movies.db
23 rows affected.
23 rows affected.
Done.
9 rows affected.
9 rows affected.
Done.
153 rows affected.
153 rows affected.
Done.
131 rows affected.
131 rows affected.
Done.
162 rows affected.
162 rows affected.


[]

3. **Check your work.**  
There are 9 theaters, 23 movies, 153 artists, 131 shows, and 162 credits.

YOUR CODE HERE

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [7]:
%%sql
SELECT MTitle, TName, Showtime, CCode, CName, COUNT(*)
FROM DATASET
GROUP BY MTitle, TName, Showtime, CCode, CName
HAVING COUNT(*) > 1

 * sqlite:///Movies.db
Done.


[('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Greg Kinnear', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Hector Elizondo', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Jon Seda', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Laurie Metcalf', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Maria Pitillo', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Roscoe Lee Browne', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'A', 'Tim Conway', 2),
 ('Dear God', 'SOCAL Canyon Springs Cinema', '9:00 PM', 'D', 'Garry Marshall', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Jon Lovitz', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Mekhi Phifer', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'Tia Carrera', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'A', 'and Louise Fletcher', 2),
 ('High School High', 'United Artists Riverside (Galleria) Tyler Mall', '9:00 PM', 'D', 'Hart Bochner', 2)]

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

In [8]:
%%sql 
DROP TABLE IF EXISTS DATASET

 * sqlite:///Movies.db
Done.


[]

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".